Begin implementing cross-site window.open in a new process

achristensen07 opened the pull request:
#### 9bcdeaaf75ed5a2cea218d8ed06fe31a3b822dc0
<pre>
Begin implementing cross-site window.open in a new process
<a href="https://bugs.webkit.org/show_bug.cgi?id=258231">https://bugs.webkit.org/show_bug.cgi?id=258231</a>
rdar://110931870

Reviewed by Chris Dumez.

This takes some ideas from <a href="https://github.com/WebKit/WebKit/pull/10169">https://github.com/WebKit/WebKit/pull/10169</a>
and some of Pascoe&apos;s more recent work, rebases on top of significant site isolation
architecture changes since then, and cleans them up with a slightly expanded test.

The back/forward navigations in the existing API test ProcessSwap.SameSiteWindowWithOpenerNavigateToFile
cover some transitions that have not been implemented yet, so I temporarily disable the test until
those transitions are implemented.  This greatly simplified this PR and allowed me to break the development
into more manageable pieces, each of which can be merged with test coverage showing progress, and each
of which is covered by a currently off-by-default feature flag.

Two important transitions need to happen when implementing window.open in a new process.  The first is
when the ProvisionalPageProxy is constructed.  The main frame already exists in the opener process
(and it is showing about:blank unless it has had a same-site URL loaded in it) but when we make a frame
in the new process we need that main frame to have an opener frame as a RemoteFrame, so we need to have
the opener inject a WebPage with a remote frame tree into the process first.  We do this by creating
a RemotePageProxy owned by the opener in the process we are provisionally navigating to.

The second transition happens when we begin receiving HTTP body for the provisional navigation and the
navigation commits.  In ProvisionalPageProxy::didCommitLoadForFrame we need to tell the frames of the
opened page in the opener process to transition to remote frames (which are still needed to be the targets
of operations like postMessage) and we do this by sending the WebPage::DidCommitLoadInAnotherProcess message.
We then create a RemotePageProxy for those remote frames to communicate with and set up message receivers.
Because the process already has a WebPage and frame tree in it, we separate the new function
RemotePageProxy::injectPageIntoNewProcess from what used to be done in each RemotePageProxy constructor.
We need to prevent this WebPage from being closed or suspended, which we do in the changes in WebPageProxy.cpp.

Since each Page can have 0 or 1 openers, we add a new member to WebPageProxy::Internals named
remotePageProxyInOpenerProcess to retain ownership of the RemotePageProxy communicating with frames
in that process.  Each Page can have an unlimited number of Pages it has opened, so the symmetric other
side of that member is a set of openedRemotePageProxies.

A few other small changes are needed to hook up remote openers, and make it possible for a main frame to
transition between local and remote, which requires changing the main frame of a WebCore::Page.

* Source/WebCore/bindings/js/WindowProxy.cpp:
(WebCore::WindowProxy::replaceFrame):
* Source/WebCore/bindings/js/WindowProxy.h:
* Source/WebCore/page/Frame.cpp:
(WebCore::Frame::resetWindowProxy):
(WebCore::Frame::takeWindowProxy):
* Source/WebCore/page/Frame.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::setMainFrame):
* Source/WebCore/page/Page.h:
* Source/WebCore/page/RemoteFrame.cpp:
(WebCore::RemoteFrame::createSubframeWithContentsInAnotherProcess):
* Source/WebCore/page/RemoteFrame.h:
* Source/WebKit/Shared/FrameInfoData.h:
* Source/WebKit/Shared/FrameInfoData.serialization.in:
* Source/WebKit/Shared/WebPageCreationParameters.cpp:
(WebKit::WebPageCreationParameters::encode const):
(WebKit::WebPageCreationParameters::decode):
* Source/WebKit/Shared/WebPageCreationParameters.h:
* Source/WebKit/UIProcess/API/APIFrameInfo.h:
* Source/WebKit/UIProcess/API/APIFrameTreeNode.h:
* Source/WebKit/UIProcess/API/Cocoa/WKFrameInfo.mm:
(-[WKFrameInfo _processIdentifier]):
* Source/WebKit/UIProcess/API/Cocoa/_WKFrameTreeNode.mm:
(-[_WKFrameTreeNode _processIdentifier]):
* Source/WebKit/UIProcess/FrameLoadState.cpp:
(WebKit::FrameLoadState::didStartProvisionalLoad):
* Source/WebKit/UIProcess/ProvisionalFrameProxy.cpp:
(WebKit::ProvisionalFrameProxy::ProvisionalFrameProxy):
* Source/WebKit/UIProcess/ProvisionalPageProxy.cpp:
(WebKit::ProvisionalPageProxy::initializeWebPage):
(WebKit::ProvisionalPageProxy::didCreateMainFrame):
(WebKit::ProvisionalPageProxy::didCommitLoadForFrame):
* Source/WebKit/UIProcess/RemotePageProxy.cpp:
(WebKit::RemotePageProxy::RemotePageProxy):
(WebKit::RemotePageProxy::injectPageIntoNewProcess):
(WebKit::RemotePageProxy::~RemotePageProxy):
* Source/WebKit/UIProcess/RemotePageProxy.h:
(WebKit::RemotePageProxy::pageID const):
(WebKit::RemotePageProxy::domain const):
* Source/WebKit/UIProcess/WebFrameProxy.h:
(WebKit::WebFrameProxy::setProcess):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::suspendCurrentPageIfPossible):
(WebKit::WebPageProxy::commitProvisionalPage):
(WebKit::WebPageProxy::shouldClosePreviousPage):
(WebKit::WebPageProxy::continueNavigationInNewProcess):
(WebKit::WebPageProxy::getAllFrameTrees):
(WebKit::WebPageProxy::createNewPage):
(WebKit::WebPageProxy::addRemotePageProxy):
(WebKit::WebPageProxy::setRemotePageProxyInOpenerProcess):
(WebKit::WebPageProxy::addOpenedRemotePageProxy):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxyInternals.h:
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::processForNavigation):
* Source/WebKit/WebProcess/WebCoreSupport/WebRemoteFrameClient.cpp:
(WebKit::WebRemoteFrameClient::changeLocation):
* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
(WebKit::WebFrame::info const):
(WebKit::WebFrame::didCommitLoadInAnotherProcess):
(WebKit::WebFrame::transitionToLocal):
* Source/WebKit/WebProcess/WebPage/WebFrame.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::WebPage):
(WebKit::m_appHighlightsVisible):
(WebKit::WebPage::didCommitLoadInAnotherProcess):
(WebKit::WebPage::transitionFrameToLocal):
(WebKit::WebPage::transitionFrameToLocalAndLoadRequest): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Canonical link: <a href="https://commits.webkit.org/265321@main">https://commits.webkit.org/265321@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f39ee89efe999e5789aa92223bf4b1524806bb03

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/10594 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/10816 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/11096 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/12241 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/10165 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/10609 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/13188 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/10783 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/13094 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/10756 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/11680 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/8909 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/12642 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/8974 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/9567 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/16828 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/10045 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/9717 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/12971 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/10184 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/8264 "5 failures") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/9338 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/13602 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/1181 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/10040 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/page/Frame.cpp:
```diff
@@ -104,4 +105,12 @@ void Frame::disconnectOwnerElement()
         document->frameWasDisconnectedFromOwner();
 }
 
+void Frame::takeWindowProxy(Frame& frame)
```
I find the naming slightly confusing. Maybe it should be `takeWindowProxyFrom()`.

cdumez commented on Source/WebKit/UIProcess/FrameLoadState.cpp:
```diff
@@ -46,7 +46,7 @@ void FrameLoadState::removeObserver(Observer& observer)
 void FrameLoadState::didStartProvisionalLoad(const URL& url)
 {
-    ASSERT(m_provisionalURL.isEmpty());
+    ASSERT(m_provisionalURL.isEmpty() || m_provisionalURL == url);
```
Seems like we may be papering over a real issue here? We don't want didStartProvisionalLoad to get called several times for the same load. This would be observable by the client app too.

achristensen07 replied:
I'll look closer.  You're probably right

cdumez commented on Source/WebKit/UIProcess/WebPageProxyInternals.h:
```diff
@@ -215,6 +215,8 @@ struct WebPageProxy::Internals final : WebPopupMenuProxy::Client
 
     HashMap<WebCore::RegistrableDomain, WeakPtr<RemotePageProxy>> domainToRemotePageProxyMap;
+    RefPtr<RemotePageProxy> remotePageProxyInOpenerProcess;
+    HashSet<Ref<RemotePageProxy>> openedRemotePageProxies;
```
seems like we're leaking these objects here?

achristensen07 replied:
There's a FIXME comment by WebPageProxy::addOpenedRemotePageProxy

cdumez commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ProcessSwapOnNavigation.mm:
```diff
@@ -1390,7 +1390,8 @@ static void enableWindowOpenPSON(WKWebViewConfiguration *configuration)
 }
 
-TEST(ProcessSwap, CrossSiteWindowOpenWithOpener)
+// FIXME: Get this working.
```
Doesn't seem OK to regress pre-existing PSON API tests with new feature.

Why is there even a behavior change here? Isn't the new behavior gated by a feature flag?

achristensen07 replied:
These two tests are the two existing tests that enable the ProcessSwapOnCrossSiteWindowOpenEnabled feature.  That's why they are affected.

cdumez replied:
Oh, I see. No concern then.

cdumez commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ProcessSwapOnNavigation.mm:
```diff
@@ -7178,7 +7179,8 @@ static bool hasOverlay(CALayer *layer)
 static const char* openedPage = "Hello World";
 
-TEST(ProcessSwap, SameSiteWindowWithOpenerNavigateToFile)
+// FIXME: Get this working.
```
Doesn't seem OK to regress pre-existing PSON API tests with new feature.

cdumez reviewed (approved):
LGTM though it'd be nice to fix the duplicate calls to didStartProvisionalLoad.

achristensen07 committed:
Begin implementing cross-site window.open in a new process
https://bugs.webkit.org/show_bug.cgi?id=258231
rdar://110931870

Reviewed by Chris Dumez.

This takes some ideas from https://github.com/WebKit/WebKit/pull/10169
and some of Pascoe's more recent work, rebases on top of significant site isolation
architecture changes since then, and cleans them up with a slightly expanded test.

The back/forward navigations in the existing API test ProcessSwap.SameSiteWindowWithOpenerNavigateToFile
cover some transitions that have not been implemented yet, so I temporarily disable the test until
those transitions are implemented.  This greatly simplified this PR and allowed me to break the development
into more manageable pieces, each of which can be merged with test coverage showing progress, and each
of which is covered by a currently off-by-default feature flag.

Two important transitions need to happen when implementing window.open in a new process.  The first is
when the ProvisionalPageProxy is constructed.  The main frame already exists in the opener process
(and it is showing about:blank unless it has had a same-site URL loaded in it) but when we make a frame
in the new process we need that main frame to have an opener frame as a RemoteFrame, so we need to have
the opener inject a WebPage with a remote frame tree into the process first.  We do this by creating
a RemotePageProxy owned by the opener in the process we are provisionally navigating to.

The second transition happens when we begin receiving HTTP body for the provisional navigation and the
navigation commits.  In ProvisionalPageProxy::didCommitLoadForFrame we need to tell the frames of the
opened page in the opener process to transition to remote frames (which are still needed to be the targets
of operations like postMessage) and we do this by sending the WebPage::DidCommitLoadInAnotherProcess message.
We then create a RemotePageProxy for those remote frames to communicate with and set up message receivers.
Because the process already has a WebPage and frame tree in it, we separate the new function
RemotePageProxy::injectPageIntoNewProcess from what used to be done in each RemotePageProxy constructor.
We need to prevent this WebPage from being closed or suspended, which we do in the changes in WebPageProxy.cpp.

Since each Page can have 0 or 1 openers, we add a new member to WebPageProxy::Internals named
remotePageProxyInOpenerProcess to retain ownership of the RemotePageProxy communicating with frames
in that process.  Each Page can have an unlimited number of Pages it has opened, so the symmetric other
side of that member is a set of openedRemotePageProxies.

A few other small changes are needed to hook up remote openers, and make it possible for a main frame to
transition between local and remote, which requires changing the main frame of a WebCore::Page.

* Source/WebCore/bindings/js/WindowProxy.cpp:
(WebCore::WindowProxy::replaceFrame):
* Source/WebCore/bindings/js/WindowProxy.h:
* Source/WebCore/page/Frame.cpp:
(WebCore::Frame::resetWindowProxy):
(WebCore::Frame::takeWindowProxy):
* Source/WebCore/page/Frame.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::setMainFrame):
* Source/WebCore/page/Page.h:
* Source/WebCore/page/RemoteFrame.cpp:
(WebCore::RemoteFrame::createSubframeWithContentsInAnotherProcess):
* Source/WebCore/page/RemoteFrame.h:
* Source/WebKit/Shared/FrameInfoData.h:
* Source/WebKit/Shared/FrameInfoData.serialization.in:
* Source/WebKit/Shared/WebPageCreationParameters.cpp:
(WebKit::WebPageCreationParameters::encode const):
(WebKit::WebPageCreationParameters::decode):
* Source/WebKit/Shared/WebPageCreationParameters.h:
* Source/WebKit/UIProcess/API/APIFrameInfo.h:
* Source/WebKit/UIProcess/API/APIFrameTreeNode.h:
* Source/WebKit/UIProcess/API/Cocoa/WKFrameInfo.mm:
(-[WKFrameInfo _processIdentifier]):
* Source/WebKit/UIProcess/API/Cocoa/_WKFrameTreeNode.mm:
(-[_WKFrameTreeNode _processIdentifier]):
* Source/WebKit/UIProcess/FrameLoadState.cpp:
(WebKit::FrameLoadState::didStartProvisionalLoad):
* Source/WebKit/UIProcess/ProvisionalFrameProxy.cpp:
(WebKit::ProvisionalFrameProxy::ProvisionalFrameProxy):
* Source/WebKit/UIProcess/ProvisionalPageProxy.cpp:
(WebKit::ProvisionalPageProxy::initializeWebPage):
(WebKit::ProvisionalPageProxy::didCreateMainFrame):
(WebKit::ProvisionalPageProxy::didCommitLoadForFrame):
* Source/WebKit/UIProcess/RemotePageProxy.cpp:
(WebKit::RemotePageProxy::RemotePageProxy):
(WebKit::RemotePageProxy::injectPageIntoNewProcess):
(WebKit::RemotePageProxy::~RemotePageProxy):
* Source/WebKit/UIProcess/RemotePageProxy.h:
(WebKit::RemotePageProxy::pageID const):
(WebKit::RemotePageProxy::domain const):
* Source/WebKit/UIProcess/WebFrameProxy.h:
(WebKit::WebFrameProxy::setProcess):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::suspendCurrentPageIfPossible):
(WebKit::WebPageProxy::commitProvisionalPage):
(WebKit::WebPageProxy::shouldClosePreviousPage):
(WebKit::WebPageProxy::continueNavigationInNewProcess):
(WebKit::WebPageProxy::getAllFrameTrees):
(WebKit::WebPageProxy::createNewPage):
(WebKit::WebPageProxy::addRemotePageProxy):
(WebKit::WebPageProxy::setRemotePageProxyInOpenerProcess):
(WebKit::WebPageProxy::addOpenedRemotePageProxy):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxyInternals.h:
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::processForNavigation):
* Source/WebKit/WebProcess/WebCoreSupport/WebRemoteFrameClient.cpp:
(WebKit::WebRemoteFrameClient::changeLocation):
* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
(WebKit::WebFrame::info const):
(WebKit::WebFrame::didCommitLoadInAnotherProcess):
(WebKit::WebFrame::transitionToLocal):
* Source/WebKit/WebProcess/WebPage/WebFrame.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::WebPage):
(WebKit::m_appHighlightsVisible):
(WebKit::WebPage::didCommitLoadInAnotherProcess):
(WebKit::WebPage::transitionFrameToLocal):
(WebKit::WebPage::transitionFrameToLocalAndLoadRequest): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Canonical link: https://commits.webkit.org/265321@main
